### PR TITLE
Sharding: copy primary key tables only at the end of the ferry

### PR DESCRIPTION
**Pull request by hkdsun**
This change helps keep the primary key table in the target database consistent for a longer time period.

For example, for us, we have developers who `JOIN ON` a primary key table in their queries to filter out some data but see unexpected results. This change makes that sort of inconvenience less likely

follow up to: https://github.com/Shopify/shopify/pull/153139
related to: https://github.com/Shopify/pods/issues/259
@Shopify/pods

### CODE REVIEW

**Comment by hkdsun** [comment]
Added table verification logic which required further refactor of the iterative verifier.
 
Found a bug in the `Pool` helper on the way where we were not consuming the whole error channel as well.

@fw42 could you have another look?

File: utils.go [hunk 1]
```diff
@@ -115,12 +115,27 @@ loop:
...
 		case workQueue <- i:
 			i++
-		case err = <-errCh:
-			break loop
+		case err = <-errCh: // abort pool if an error was discovered
+			if err != nil {
+				break loop
+			}
+		default:
+			time.Sleep(500 * time.Millisecond)
```
Comment (hkdsun) [inline utils.go]: We were busy-looping previously if all goroutines were busy doing work

File: sharding/sharding.go [hunk 2]
```diff
@@ -212,6 +216,59 @@ func (r *ShardingFerry) deltaCopyJoinedTables() error {
...
+		if _, exists := pkTables[table.Name]; exists {
+			if len(table.PKColumns) != 1 {
+				return fmt.Errorf("Multiple PK columns are not supported with the PrimaryKeyTables option")
+			}
+			tables = append(tables, table)
+		}
+	}
+
+	if len(tables) == 0 {
+		r.logger.Warn("found no primary key tables to copy")
```
Comment (fw42) [inline sharding/sharding.go]: I assume this is for cases where the "shops table isn't sharded yet"?

Comment (hkdsun) [reply sharding/sharding.go]: yeah or the PrimaryKeyTables config is not provided. maybe we should return an error here

File: utils.go [hunk 3]
```diff
@@ -115,12 +115,27 @@ loop:
...
 
 	close(workQueue)
 	wg.Wait()
+	close(errCh)
+
+	if err != nil {
+		return results, err
+	}
+
+	for e := range errCh {
```
Comment (fw42) [inline utils.go]: If there's multiple errors in the channel, this will make it so that we return (only) the last one. That's intentional?

Comment (hkdsun) [reply utils.go]: yeah. unless we return an array of errors which is unconventional. what would you expect?

File: sharding/sharding.go (hunk 2 above)
Comment (fw42) [reply sharding/sharding.go]: If we return an error then people who's tenant table isn't sharded can't use this, right? I think we should allow them to.

Comment (fw42) [reply sharding/sharding.go]: Or do you mean return an error if a table name is specified but the table isn't found? If so then I agree.

Comment (hkdsun) [reply sharding/sharding.go]: yup latter is what i meant. will push that change

File: utils.go (hunk 3 above)
Comment (fw42) [reply utils.go]: I'd probably expect to see the first instead of the last error, but no big deal.

**Comment by hkdsun** [comment]
Shuhao and Justin are both away for a few days so I'm going to merge this. Feel free to leave comments on this PR and I'll come back to them

**Pull request merged**